readers-frontend/src/pages/[year]/[month]/[day]/[slug].astro

```diff
@@ -0,0 +1,326 @@
+---
+import Article from "@components/Article.astro";
+import ArticleCookieManager from "@components/ArticleCookieManager.astro";
+import TopArticles from "@components/SolidJS/TopArticles";
+import {
+  articleWithUserAndDraft,
+  articlesWithArticleDrafts,
+  getAllArticles,
+  getArticleImage,
+  getArticles,
+} from "@components/utils/articles";
+import type { articlesCategories, categories } from "@prisma/client";
+import { Icon } from "astro-icon";
+import Layout from "src/layouts/Layout.astro";
+import prisma from "src/prisma";
+
+type ArticleCategories = articlesCategories & {
+  category: categories;
+};
+
+export const getParentCategory = (categories: ArticleCategories[]) => {
+  // I think categories commonly come from the backend as:
+  // Top - Middle - Bottom for examples. Where Bottom is nested
+  // under Middle and Top is an "extra category". Hence, the reverse,
+  // trying to find related ones.
+  const interimParentCategory = categories
+    .slice()
+    .reverse()
+    .find(
+      ({ category }) =>
+        (category.categories_nestUnder === null && // Nouse
+          category.categories_id !== 4) ||
+        category.categories_nestUnder === 4 // Muse
+    );
+
+  if (interimParentCategory) {
+    if (
+      // This is mainly for Muse articles as they are
+      // always going to have the Muse category listed
+      categories.length < 3 ||
+      categories.find(
+        ({ category }) =>
+          category.categories_nestUnder ===
+          interimParentCategory.category.categories_id
+      )
+    )
+      return interimParentCategory.category;
+  }
+
+  return categories[0].category;
+};
+
+export const dateFormatter = new Intl.DateTimeFormat("en-GB", {
+  year: "numeric",
+  month: "2-digit",
+  day: "2-digit",
+});
+
+export async function getStaticPaths() {
+  const articles = (await getAllArticles()).filter((article) => {
+    // Ensure article has an articleDraft
+    if (article.articlesDrafts.length === 0) return false;
+
+    // Ensure article has categories
+    if (article.categories.length === 0) return false;
+
+    // Some articles' slug is not formatted properly
+    const slug = article.articles_slug;
+    if (slug)
+      return (
+        !slug.includes("/") &&
+        !slug.includes("\\") &&
+        !slug.includes("#") &&
+        !slug.includes("?")
+      );
+
+    return false;
+  });
+
+  const similarArticles = await articles.reduce(
+    async (accumulatorPromise, article) => {
+      const accumulator = await accumulatorPromise;
+
+      const parentCategoryId = getParentCategory(
+        article.categories
+      ).categories_id;
+
+      // Get 4 most recent articles in parent category
+      const mSimilarArticlesIds = (
```

Comment:
Select with only one column is quick regardless, so the time spent on this is negligible.

From the docker container:
```
bash-4.4# cat select-test.sql 
SELECT articles.articles_id FROM articles WHERE articles.articles_categories LIKE '%373%' AND articles.articles_id != 148297 ORDER BY articles.articles_published DESC LIMIT 4;
bash-4.4# time echo $(cat select-test.sql) | mysql -u userDocker -p nouse
Enter password: 
articles_id
151069
151059
151058
151057

real    0m2.354s
user    0m0.005s
sys     0m0.008s
bash-4.4#
```

I don't want to put the include here cause we have the `getArticles` function, which does just that.

Reply:
It's not unnecessary. The extra operation is to use the `getArticles` function as it's meant to be used. It wants a list of article ids, so those ids need to come from somewhere. Using `include` does not make sense when there's a function that already does that. Not sure what the database service would address here?

Reply:
You're saying that adding include does not make sense when `getArticles` already does it but `getArticles` is only supposed to grab by the IDs. You're already adding an extra `findMany` operation just to grab the 4 IDs. It's unnecessary overhead when the added findMany operation can just do it all at once yk? As for the database service, it might address this because the articles would already be in memory so no need for multiple SQL operations

Reply:
It's 2ms overhead. What's the point of having the function if it's not reusable?